cheri: Fix qsort to be cap-aware

heshamelmatary opened the pull request:
This is imitated from https://github.com/CTSRD-CHERI/cheribsd/blob/master/lib/libc/stdlib/qsort.c

arichardson commented on newlib/libc/search/qsort.c:
```diff
@@ -70,6 +70,12 @@ PORTABILITY
 #define inline
 #endif
 
+#ifdef __CHERI__
```
We generally use `#if __has_feature(capabilities)` nowadays, but maybe newlib doesn't have the non-clang fallback defined? In that case #fidef __CHERI__ is fine.

heshamelmatary replied:
Yeah, I also followed the convention in mem*() family in newlib; they use `#ifdef __CHERI__`

jrtc27 replied:
Most of the newlib code dates from when we still had a lot of `__CHERI__` in CheriBSD (or at least if it was changing I was not yet aware of that) so I'm to blame for the ones that are already there. newlib's sys/cdefs.h does have a `__has_feature` fallback as upstream actually uses `__has_feature(nullability)`, and both cheric.h and sys/_stdint.h use `__has_feature(capabilities)`.

heshamelmatary replied:
I changed that to use `__has_feature(capabilities)`

jrtc27 commented on newlib/libc/search/qsort.c:
```diff
@@ -96,30 +102,30 @@ static inline void	 swapfunc (char *, char *, int, int);
+		*(big_primitive_type *)(a) = *(big_primitive_type *)(b);  \
+		*(big_primitive_type *)(b) = t;  \
+	} else  \
+		swapfunc(a, b, es, swaptype_big_primitive_type)
```
Please keep the upstream formatting

jrtc27 commented on newlib/libc/search/qsort.c:
```diff
@@ -187,7 +193,7 @@ qsort (void *a,
 	size_t d, r;
 	int cmp_result;
-	int swaptype, swap_cnt;
+	int swaptype_big_primitive_type, swap_cnt;
```
Why does this need renaming?

heshamelmatary replied:
The macros now use that variable name (`swaptype_big_primitive_type`) which is more indicative?

jrtc27 replied:
Not really. It's only called that in FreeBSD because there's also a `swaptype_int`. This just unnecessarily adds to the diff for purely cosmetic reasons (that I don't think are even good).

heshamelmatary committed:
cheri: Fix qsort to be cap-aware

This is imitated from https://github.com/CTSRD-CHERI/cheribsd/blob/master/lib/libc/stdlib/qsort.c

arichardson reviewed (approved):
Seems fine to me.
